app/src/main/java/org/thoughtcrime/securesms/notifications/MarkReadReceiver.kt

```diff
@@ -0,0 +1,120 @@
+import android.content.Intent
+import android.os.AsyncTask
+import androidx.core.app.NotificationManagerCompat
+import com.annimon.stream.Collectors
+import com.annimon.stream.Stream
+import nl.komponents.kovenant.task
+import org.session.libsession.messaging.MessagingModuleConfiguration.Companion.shared
+import org.session.libsession.messaging.messages.control.ReadReceipt
+import org.session.libsession.messaging.sending_receiving.MessageSender.send
+import org.session.libsession.snode.SnodeAPI
+import org.session.libsession.snode.SnodeAPI.nowWithOffset
+import org.session.libsession.utilities.TextSecurePreferences
+import org.session.libsession.utilities.TextSecurePreferences.Companion.isReadReceiptsEnabled
+import org.session.libsession.utilities.associateByNotNull
+import org.session.libsession.utilities.recipients.Recipient
+import org.session.libsignal.utilities.Log
+import org.thoughtcrime.securesms.ApplicationContext
+import org.thoughtcrime.securesms.database.MessagingDatabase.ExpirationInfo
+import org.thoughtcrime.securesms.database.MessagingDatabase.MarkedMessageInfo
+import org.thoughtcrime.securesms.database.MessagingDatabase.SyncMessageId
+import org.thoughtcrime.securesms.dependencies.DatabaseComponent
+import org.thoughtcrime.securesms.util.SessionMetaProtocol.shouldSendReadReceipt
+
+class MarkReadReceiver : BroadcastReceiver() {
+    @SuppressLint("StaticFieldLeak")
+    override fun onReceive(context: Context, intent: Intent) {
+        if (CLEAR_ACTION != intent.action) return
+        val threadIds = intent.getLongArrayExtra(THREAD_IDS_EXTRA)
+        if (threadIds != null) {
+            NotificationManagerCompat.from(context)
+                .cancel(intent.getIntExtra(NOTIFICATION_ID_EXTRA, -1))
+            object : AsyncTask<Void?, Void?, Void?>() {
+                override fun doInBackground(vararg params: Void?): Void? {
+                    val currentTime = nowWithOffset
+                    for (threadId in threadIds) {
+                        Log.i(TAG, "Marking as read: $threadId")
+                        val storage = shared.storage
+                        storage.markConversationAsRead(threadId, currentTime, true)
+                    }
+                    return null
+                }
+            }.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR)
+        }
+    }
+
+    companion object {
+        private val TAG = MarkReadReceiver::class.java.getSimpleName()
+        const val CLEAR_ACTION = "network.loki.securesms.notifications.CLEAR"
+        const val THREAD_IDS_EXTRA = "thread_ids"
+        const val NOTIFICATION_ID_EXTRA = "notification_id"
+
+        @JvmStatic
+        fun process(context: Context, markedReadMessages: List<MarkedMessageInfo>) {
+
+            val loki = DatabaseComponent.get(context).lokiMessageDatabase()
+
+            task {
+                val hashToInfo = markedReadMessages.associateByNotNull { loki.getMessageServerHash(it.expirationInfo.id) }
```

Comment:
We might still have to perform a migration for the message server hash stuff, to include (and pass in here) whether the hashes are from MMS or SMS

Reply:
it would basically be in `SQLCipherOpenHelper`, updating the current version, adding a new version one above the current latest `lokiVwhatever = 45` or something... and then adding a column to the `LokiMessageDatabase` which is, I guess INT DEFAULT 0 called `mms` or whichever the normal default is I guess.. I think whether it's `mms` would follow the convention in `MessagingDatabase` I guess (not specifying the column to be sms = 1 or 0), then might have to update where the server hashes are being store to include that and add it to the content values, it might only be like two places I guess where we might be adding it. the get side might only be a small handful as well... tag me if this doesn't make any sense :sweat_smile:

Reply:
should be sorted now with the new tables.